### PR TITLE
ALC Period Info Widget

### LINKED ISSUE #31798: ALC Muon Period Info Widget
This involves rewriting the Muon Period Info widget in C++ to allow it to be used in the ALC and then exposing it to python for use in Muon Analysis

### PR DESCRIPTION
**Pull request by Matt-Cumber**
**Description of work.**
Last release a new table widget was added to the muon interfaces which displays period information from the currently loaded run. This PR implements the same table for the ALC interface too. Since ALC is written in C++ and MA/FDA in python, the widget was rewritten in C++ along with it's tests and uses sip to export the widget for use in the python GUI's

**To test:**
Open ALC interface
Load HIFI 110542
Click the periods button
There should be 8 periods with a table containing data similar to this (taken from Wimda)
image
The choice of columns was already decided in the previous [PR](https://github.com/mantidproject/mantid/pull/30939)
The same table should be reproducible in Muon Analysis and Frequency Domain Analysis.
Go to the grouping tab in both interface and click the periods button
Do some stress testing such as clicking the button with no data, load new data with the table open, clear the MA/FDA Guis with it open etc.

Fixes #31798 

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: qt/widgets/common/inc/MantidQtWidgets/Common/MuonPeriodInfo.h [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+                        const std::string &totalFrames, const std::string &counts, const std::string &tag);
+  /// Gets all sample log data related to periods
+  std::vector<std::vector<std::string>> getInfo(Mantid::API::MatrixWorkspace_sptr ws);
+  /// Takes the workspace and adds it's period info to the table if any
+  void addInfo(const Mantid::API::Workspace_sptr ws);
+  /// Set the title of the widget
+  void setWidgetTitleRuns(const std::string &title);
+  /// Get the title of the widget
+  std::string getWidgetTitleRuns() const;
+  /// Set the number of sequences that was gathered
```
Comment (AnthonyLim23) [inline qt/widgets/common/inc/MantidQtWidgets/Common/MuonPeriodInfo.h:42]: what do you mean by sequence?

Comment (Matt-Cumber) [reply qt/widgets/common/inc/MantidQtWidgets/Common/MuonPeriodInfo.h:42]: To be honest this is just the name given in the nexus file, I am not 100% sure what is meant by it

Comment (AnthonyLim23) [reply qt/widgets/common/inc/MantidQtWidgets/Common/MuonPeriodInfo.h:42]: could you ask Steve?

File: qt/widgets/common/test/MuonPeriodInfoTest.h [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+  }
+
+  void test_DAQ_count_increases_as_expected() {
+    m_periodInfo->addPeriodToTable("state 1", "1", "10", "200", "25", "1");
+    TS_ASSERT_EQUALS(1, m_periodInfo->getDAQCount());
+    m_periodInfo->addPeriodToTable("state 1 dwell", "2", "10", "200", "25", "1");
+    TS_ASSERT_EQUALS(1, m_periodInfo->getDAQCount());
+    m_periodInfo->addPeriodToTable("state 1", "1", "10", "200", "25", "1");
+    m_periodInfo->addPeriodToTable("state 1", "1", "10", "200", "25", "1");
+    TS_ASSERT_EQUALS(3, m_periodInfo->getDAQCount());
```
Comment (AnthonyLim23) [inline qt/widgets/common/test/MuonPeriodInfoTest.h:71]: what trigged the DAQ count to increase?

File: qt/widgets/common/test/MuonPeriodInfoTest.h [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+  }
+  void test_setNumberOfSequences_to_negative() {
+    m_periodInfo->setNumberOfSequences(-1);
+    TS_ASSERT_EQUALS("Number of period cycles not found", m_periodInfo->getNumberOfSequencesString());
+  }
+
+  void test_setNumberOfSequences() {
+    m_periodInfo->setNumberOfSequences(2);
+    TS_ASSERT_EQUALS("Run contains 2 cycles of periods", m_periodInfo->getNumberOfSequencesString());
```
Comment (AnthonyLim23) [inline qt/widgets/common/test/MuonPeriodInfoTest.h:80]: does this mean that the sequences are the number of periods?

File: qt/widgets/common/test/MuonPeriodInfoTest.h (hunk 2 above)
Comment (Matt-Cumber) [reply qt/widgets/common/test/MuonPeriodInfoTest.h:71]: Adding a DAQ period to the table. the second argument is the type of period (1=DAQ, 2=DWELL). Here we added two daq periods as they both have the second argument as 1

File: qt/widgets/common/test/MuonPeriodInfoTest.h (hunk 3 above)
Comment (Matt-Cumber) [reply qt/widgets/common/test/MuonPeriodInfoTest.h:80]: No they are not 2 is just a random arbitrary number I used for the test. This text format is the same as Wimbda

File: qt/widgets/common/test/MuonPeriodInfoTest.h (hunk 2 above)
Comment (AnthonyLim23) [reply qt/widgets/common/test/MuonPeriodInfoTest.h:71]: Could we have comment explaining that?

File: qt/widgets/common/test/MuonPeriodInfoTest.h (hunk 3 above)
Comment (AnthonyLim23) [reply qt/widgets/common/test/MuonPeriodInfoTest.h:80]: ok. I just want to know what it means

File: qt/widgets/common/inc/MantidQtWidgets/Common/MuonPeriodInfo.h (hunk 1 above)
Comment (Matt-Cumber) [reply qt/widgets/common/inc/MantidQtWidgets/Common/MuonPeriodInfo.h:42]: From Steve:
_The entry period_sequences records how many times the complete period sequence was executed during the run. For instance, if we look at EMU run 114295, which was a run carried out with a two period sequence collecting 500 frames in each period (i.e. 1000 frames to complete a single sequence), period_sequences reads 218 suggesting 218000 frames have been collected during the run. This compares well with frames_good (218002), a record of the total number of frames collected. (not sure why we’ve recorded two additional frames in the final period, but that’s another matter)._

Comment (AnthonyLim23) [reply qt/widgets/common/inc/MantidQtWidgets/Common/MuonPeriodInfo.h:42]: So in simple terms: a sequence is completing recording for all of the periods once?

Can we add a comment explaining this?

**Comment by AnthonyLim23** [comment]
Opening FDA causes a crash

**Comment by AnthonyLim23** [comment]
When I open the periods window in MA or FDA, it sort of merges with the main Mantid tool bar. So I cannot move it.

**Comment by AnthonyLim23** [comment]
image

**Review by AnthonyLim23 — APPROVED** [review]
It all works now as expected. :shipit:

**Pull request merged**